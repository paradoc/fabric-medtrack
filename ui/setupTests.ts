import '@testing-library/jest-dom/extend-expect'


import dayjs from 'dayjs'
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
import duration from 'dayjs/plugin/duration'
import relativeTime from 'dayjs/plugin/relativeTime'

dayjs.extend(isSameOrAfter)
dayjs.extend(duration)
dayjs.extend(relativeTime)


export function setupFetchStub(data: any) {
  return function fetchStub(_url: string) {
    return new Promise((resolve) => {
      resolve({
        json: () => Promise.resolve(data),
      })
    })
  }
}