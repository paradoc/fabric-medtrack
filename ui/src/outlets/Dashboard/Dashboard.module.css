.container {
  display: grid;
  grid-template-rows: min-content 1fr;
  row-gap: 56px;
}

.controls {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  column-gap: 48px;
}

.controls > a {
  display: grid;
  text-decoration: none;
}

.control {
  height: 192px;
  border: none;
  border-radius: 12px;
  font-weight: bold;
  text-transform: uppercase;
  font-size: 16px;
  color: #444;
}

.feed {
  display: grid;
  grid-template-rows: min-content min-content;
  row-gap: 24px;
}

.feed > header {
  font-size: 28px;
  font-weight: bold;
  color: #333;
}

.feedList {
  display: grid;
  row-gap: 24px;  
}

.feedItem {
  display: grid;
  grid-template-columns: min-content max-content max-content;
  column-gap: 24px;
  color: #333;
}

.dispatchId,
.time {
  font-weight: 100;
}
