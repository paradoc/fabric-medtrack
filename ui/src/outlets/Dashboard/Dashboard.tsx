import React, { useCallback, useEffect, useState } from 'react'
import { Link } from 'react-router-dom'
import { prop } from 'rambda'

import { DispatchData } from '../../views/Collector'

import styles from './Dashboard.module.css'

function Feed() {
  // TODO: Get latest N dispatches from backend
  const [feed, setFeed] = useState<DispatchData[]>([])

  const fetchData = useCallback(async () => {
    const response = await fetch('/api/recent/7')
    const data = await response.json()
    setFeed(data)
  }, [])

  useEffect(() => {
    const id = setInterval(fetchData, 2000)
    return () => {
      clearInterval(id)
    }
  }, [])

  return (
    <section className={styles.feed}>
      <header>Dispatch Feed</header>
      <div className={styles.feedList}>
        {feed.map(({ dispatch_id, dispatch_date, medications }) => (
          <div className={styles.feedItem} key={dispatch_id}>
            <span className={styles.time}>{dispatch_date}</span>
            <span className={styles.meds}>
              {medications.map(prop('generic_name')).join(', ')}
            </span>
            <Link to={`./inspect/${dispatch_id}`} relative="path">
              <span className={styles.dispatchId}>{dispatch_id}</span>
            </Link>
          </div>
        ))}
      </div>
    </section>
  )
}

export default function Dashboard() {
  return (
    <div className={styles.container}>
      <section className={styles.controls}>
        <Link to="/dispatcher/new">
          <button className={styles.control}>dispatch</button>
        </Link>

        {/** We don't have a plan to implement this for now. */}
        <button className={styles.control}>
          inspect
          <br />
          (non-functional)
        </button>
        <button className={styles.control}>
          history
          <br />
          (non-functional)
        </button>
      </section>
      <Feed />
    </div>
  )
}
