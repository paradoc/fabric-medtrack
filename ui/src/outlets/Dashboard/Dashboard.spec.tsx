import React from 'react'
import { MemoryRouter } from 'react-router'
import { act } from 'react-dom/test-utils'
import { render, waitFor } from '@testing-library/react'
import Dashboard from './Dashboard'

import { setupFetchStub } from '../../../setupTests'

describe('Dashboard', () => {
  beforeEach(() => {
    const fakeData = [
      {
        dispatch_id: 'test-dispatch-id',
        medications: [
          { generic_name: 'test-generic', brand_name: 'test_brand' },
        ],
      },
    ]
    global.fetch = jest.fn().mockImplementation(setupFetchStub(fakeData))
  })

  afterEach(() => {
    // global.fetch.mockClear()
    // delete global.fetch
    jest.clearAllTimers()
  })

  test('should match snapshot', async () => {
    jest.useFakeTimers()
    jest.runAllTimers()
    const { container, getByText } = render(
      <MemoryRouter initialEntries={['/']}>
        <Dashboard />
      </MemoryRouter>
    )
    await waitFor(
      () => {
        expect(getByText('test-dispatch-id')).toBeInTheDocument()
      },
      {
        timeout: 5000,
        interval: 2000,
      }
    )
    expect(container).toMatchSnapshot()
  })
})
