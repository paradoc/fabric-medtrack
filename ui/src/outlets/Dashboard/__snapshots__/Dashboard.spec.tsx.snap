// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <section
      class="controls"
    >
      <a
        href="/dispatcher/new"
      >
        <button
          class="control"
        >
          dispatch
        </button>
      </a>
      <button
        class="control"
      >
        inspect
        <br />
        (non-functional)
      </button>
      <button
        class="control"
      >
        history
        <br />
        (non-functional)
      </button>
    </section>
    <section
      class="feed"
    >
      <header>
        Dispatch Feed
      </header>
      <div
        class="feedList"
        data-testid="feeds"
      >
        <div
          class="feedItem"
        >
          <span
            class="time"
          />
          <span
            class="meds"
          >
            test-generic
          </span>
          <a
            href="/inspect/test-dispatch-id"
          >
            <span
              class="dispatchId"
            >
              test-dispatch-id
            </span>
          </a>
        </div>
      </div>
    </section>
  </div>
</div>
`;
